File: tests/oracle/infertypes_settle_1204.cc
/* INFERTYPES-SETTLE-0001: ActionInferTypes interning/settling fixture.
 *
 * Exercises the float-conversion chain and the LOAD/STORE pointer edges that
 * drive TypeOp::propagateToPointer/propagateFromPointer (typeop.cc:186-228),
 * then applies ActionInferTypes (coreaction.cc:5374-5416) repeatedly and
 * observes the per-round Varnode type table together with the pointer
 * identity stability against the previous round. The oracle's writeBack
 * settles because every propagated Datatype is TypeFactory-interned: a
 * stable table plus stable pointer identities from round to round is the
 * observable contract the "Type propagation algorithm not settling"
 * warning (coreaction.cc:5390-5392) depends on.
 */
#include <bits/stdc++.h>

#include "architecture.hh"
#include "capability.hh"
#include "coreaction.hh"
#include "database.hh"
#include "funcdata.hh"
#include "libdecomp.hh"
#include "marshal.hh"
#include "op.hh"
#include "translate.hh"
#include "type.hh"
#include "typeop.hh"
#include "varnode.hh"

using namespace std;
using namespace ghidra;

class FixtureTranslate final : public Translate {
  VarnodeData dummyRegister;

public:
  FixtureTranslate()
  {
    setBigEndian(false);
    setUniqueBase(0);
    insertSpace(new ConstantSpace(this, this));
    insertSpace(new AddrSpace(this, this, IPTR_PROCESSOR, "other", false,
                              8, 1, 1, AddrSpace::hasphysical, 0, 0));
    insertSpace(new UniqueSpace(this, this, 2, 0));
    AddrSpace *ram = new AddrSpace(this, this, IPTR_PROCESSOR, "ram", false,
                                   8, 1, 3, AddrSpace::hasphysical, 0, 0);
    insertSpace(ram);
    AddrSpace *reg = new AddrSpace(this, this, IPTR_PROCESSOR, "register", false,
                                   8, 1, 4, AddrSpace::hasphysical, 0, 0);
    insertSpace(reg);
    SpacebaseSpace *stack = new SpacebaseSpace(
        this, this, "stack", 5, 8, ram, 1, true);
    insertSpace(stack);
    VarnodeData stackPointer;
    stackPointer.space = reg;
    stackPointer.offset = 0;
    stackPointer.size = 8;
    addSpacebasePointer(stack, stackPointer, 8, true);
    insertSpace(new AddrSpace(this, this, IPTR_PROCESSOR, "join", false,
                              8, 1, 6, AddrSpace::hasphysical, 0, 0));
    insertSpace(new IopSpace(this, this, 7));
    setDefaultCodeSpace(3);
    dummyRegister.space = reg;
    dummyRegister.offset = 0;
    dummyRegister.size = 8;
  }

  void initialize(DocumentStorage &) override {}
  const VarnodeData &getRegister(const string &) const override {
    return dummyRegister;
  }
  string getRegisterName(AddrSpace *, uintb, int4) const override { return ""; }
  string getExactRegisterName(AddrSpace *, uintb, int4) const override { return ""; }
  void getAllRegisters(map<VarnodeData, string> &) const override {}
  void getUserOpNames(vector<string> &) const override {}
  int4 instructionLength(const Address &) const override { return 0; }
  int4 oneInstruction(PcodeEmit &, const Address &) const override { return 0; }
  int4 printAssembly(AssemblyEmit &, const Address &) const override { return 0; }
};

class FixtureArchitecture final : public Architecture {
protected:
  Translate *buildTranslator(DocumentStorage &) override { return (Translate *)0; }
  void buildLoader(DocumentStorage &) override {}
  PcodeInjectLibrary *buildPcodeInjectLibrary(void) override {
    return (PcodeInjectLibrary *)0;
  }
  void buildTypegrp(DocumentStorage &) override {}
  void buildCoreTypes(DocumentStorage &) override {}
  void buildCommentDB(DocumentStorage &) override {}
  void buildStringManager(DocumentStorage &) override {}
  void buildConstantPool(DocumentStorage &) override {}
  void buildContext(DocumentStorage &) override {}
  void buildSymbols(DocumentStorage &) override {}
  void buildSpecFile(DocumentStorage &) override {}
  void modifySpaces(Translate *) override {}
  void resolveArchitecture(void) override {}

public:
  FixtureArchitecture()
  {
    FixtureTranslate *fixtureTranslate = new FixtureTranslate();
    translate = fixtureTranslate;
    copySpaces(fixtureTranslate);
    max_basetype_size = 10;
    types = new TypeFactory(this);
    types->setupSizes();
    types->setCoreType("xunknown1", 1, TYPE_UNKNOWN, false);
    types->setCoreType("xunknown2", 2, TYPE_UNKNOWN, false);
    types->setCoreType("xunknown4", 4, TYPE_UNKNOWN, false);
    types->setCoreType("xunknown8", 8, TYPE_UNKNOWN, false);
    types->cacheCoreTypes();
    TypeOp::registerInstructions(inst, types, translate);
    symboltab = new Database(this, false);
    symboltab->attachScope(new ScopeInternal(0x101, "", this), (Scope *)0);
    ProtoModel *model = new ProtoModel(this);
    istringstream stream(
        "<prototype name=\"fixture\" extrapop=\"0\"><input/><output/></prototype>");
    XmlDecode decoder(this);
    decoder.ingestStream(stream);
    model->decode(decoder);
    protoModels[model->getName()] = model;
    setDefaultModel(model);
  }

  void printMessage(const string &) const override {}
};

class FixtureProgress final : public TypeStruct {
public:
  explicit FixtureProgress(const vector<TypeField> &fields)
  {
    name = "ProgressData";
    displayName = name;
    setFields(fields, 16, 8);
    markComplete();
  }
};

struct RoundCell {
  string label;
  Varnode *target;
  Datatype *type;
};

static string metaName(const Datatype *type)
{
  string result;
  metatype2string(type->getMetatype(), result);
  return result;
}

/* Byte-comparable projection: per-cell metatype/size plus whether the
 * Datatype pointer is identical to the previous round (the interned-instance
 * stability writeBack settles on). */
static string roundSnapshot(const vector<RoundCell> &cells)
{
  ostringstream stream;
  for (vector<RoundCell>::const_iterator cell = cells.begin();
       cell != cells.end(); ++cell) {
    if (cell != cells.begin()) stream << ',';
    stream << cell->label << ':' << metaName(cell->target->getType())
           << cell->target->getType()->getSize()
           << "/same_prev=" << (cell->target->getType() == cell->type);
  }
  return stream.str();
}

int main(void)
{
  cout << std::unitbuf;
  AttributeId::initialize();
  ElementId::initialize();
  CapabilityPoint::initializeAll();
  ArchitectureCapability::sortCapabilities();
  cout << "schema=1|fixture=INFERTYPES-SETTLE-0001"
       << "|oracle=e40ed13014025f82488b1f8f7bca566894ac376b\n";
  try {
    FixtureArchitecture architecture;
    AddrSpace *ram = architecture.getSpace(3);
    AddrSpace *reg = architecture.getSpace(4);
    Scope *global = architecture.symboltab->getGlobalScope();
    Funcdata fd("infertypes_settle", "infertypes_settle", global,
                Address(ram, 0x5000), (FunctionSymbol *)0, 0x40);
    BlockGraph &blocks = const_cast<BlockGraph &>(fd.getBasicBlocks());
    BlockBasic *block = blocks.newBlockBasic(&fd);
    fd.setBasicBlockRange(block, Address(ram, 0x5000), Address(ram, 0x5010));

    /* An 8-byte int input feeding the float chain. */
    Varnode *counterInput = fd.newVarnode(4, Address(reg, 0x100));
    counterInput = fd.setInputVarnode(counterInput);

    /* f2 = FLOAT_FLOAT2FLOAT(INT2FLOAT(counter) * const) : 4 -> 8 widening
     * (the RuleSubfloatConvert trigger shape). */
    PcodeOp *int2float = fd.newOp(1, Address(ram, 0x5000));
    fd.opSetOpcode(int2float, CPUI_FLOAT_INT2FLOAT);
    fd.opSetInput(int2float, counterInput, 0);
    Varnode *float4 = fd.newUniqueOut(4, int2float);
    fd.opInsertEnd(int2float, block);

    PcodeOp *fmul = fd.newOp(2, Address(ram, 0x5002));
    fd.opSetOpcode(fmul, CPUI_FLOAT_MULT);
    fd.opSetInput(fmul, float4, 0);
    Varnode *scale = fd.newConstant(4, 0x3f800000);
    fd.opSetInput(fmul, scale, 1);
    Varnode *product4 = fd.newUniqueOut(4, fmul);
    fd.opInsertEnd(fmul, block);

    PcodeOp *f2f = fd.newOp(1, Address(ram, 0x5004));
    fd.opSetOpcode(f2f, CPUI_FLOAT_FLOAT2FLOAT);
    fd.opSetInput(f2f, product4, 0);
    Varnode *double8 = fd.newUniqueOut(8, f2f);
    fd.opInsertEnd(f2f, block);

    /* A type-locked 8-byte struct pointer drives the pointer propagation
     * edges: LOAD (propagateFromPointer) and STORE (propagateToPointer). */
    Datatype *longType = architecture.types->getBase(8, TYPE_INT);
    vector<TypeField> fields;
    fields.push_back(TypeField(0, 0, "total", longType));
    fields.push_back(TypeField(1, 8, "prev", longType));
    FixtureProgress progress(fields);
    Datatype *progressPointer = architecture.types->getTypePointer(8, &progress, 1);

    Varnode *pointerInput = fd.newVarnode(8, Address(reg, 0x180));
    pointerInput = fd.setInputVarnode(pointerInput);
    pointerInput->updateType(progressPointer, true, false);

    PcodeOp *load = fd.newOp(2, Address(ram, 0x5006));
    fd.opSetOpcode(load, CPUI_LOAD);
    fd.opSetInput(load, fd.newConstant(4, ram->getIndex()), 0);
    fd.opSetInput(load, pointerInput, 1);
    Varnode *loaded8 = fd.newUniqueOut(8, load);
    fd.opInsertEnd(load, block);

    PcodeOp *store = fd.newOp(3, Address(ram, 0x5008));
    fd.opSetOpcode(store, CPUI_STORE);
    fd.opSetInput(store, fd.newConstant(4, ram->getIndex()), 0);
    fd.opSetInput(store, pointerInput, 1);
    fd.opSetInput(store, double8, 2);
    fd.opInsertEnd(store, block);

    vector<RoundCell> cells;
    const char *labels[] = { "counter", "float4", "product4", "double8",
                             "pointer", "loaded8" };
    Varnode *targets[] = { counterInput, float4, product4, double8,
                           pointerInput, loaded8 };
    for (int4 index = 0; index < 6; ++index) {
      cells.push_back(RoundCell{labels[index], targets[index],
                                targets[index]->getType()});
    }
    cout << "pre|types=" << roundSnapshot(cells) << '\n';

    fd.startTypeRecovery();
    ActionInferTypes action("typerecovery");
    action.reset(fd);
    for (int4 round = 1; round <= 8; ++round) {
      int4 result = action.apply(fd);
      cout << "round" << round << "|return=" << result
           << "|types=" << roundSnapshot(cells) << '\n';
      for (vector<RoundCell>::iterator cell = cells.begin();
           cell != cells.end(); ++cell)
        cell->type = cell->target->getType();
    }
  }
  catch (const LowlevelError &error) {
    cout << "exception|phase=run|what=" << error.explain << '\n';
    return 1;
  }
  catch (const exception &error) {
    cout << "exception|phase=run|what=" << error.what() << '\n';
    return 1;
  }
  return 0;
}

File: tests/oracle/infertypes_settle_1204.rs
// INFERTYPES-SETTLE-0001: ActionInferTypes interning/settling fixture (Rust
// twin of tests/oracle/infertypes_settle_1204.cc).
//
// Mirrors the oracle fixture op-for-op: the float-conversion chain feeding a
// 4->8 FLOAT_FLOAT2FLOAT widener plus the type-locked struct pointer driving
// LOAD/STORE pointer propagation edges. Applies ActionInferTypes eight times
// and prints the byte-comparable per-round type table with Arc-identity
// stability against the previous round — the settle observable the oracle
// gets from TypeFactory-interned Datatypes and that Rugra gets from
// canonicalize_temp_type routing every temp type through the factory.
//
// Rust-only regression guard (not printed, no oracle counterpart): the
// action's local_count must stay below the 7-pass "not settling" cap
// (coreaction.cc:5390-5392) for this graph.
use std::sync::{Arc, RwLock};

use rugra::action::Action;
use rugra::address::Address;
use rugra::arch::Architecture;
use rugra::coreaction::ActionInferTypes;
use rugra::funcdata::Funcdata;
use rugra::opcodes::OpCode;
use rugra::space::AddressSpace;
use rugra::type_system::datatype::{
    metatype2string, Datatype, TypeBase, TypeField, TypeMetatype, TypeStruct,
};
use rugra::type_system::typefactory::{CoreTypeFlavor, TypeFactory};
use rugra::varnode::Varnode;

type VarnodeRef = Arc<RwLock<Varnode>>;

struct RoundCell {
    label: &'static str,
    target: VarnodeRef,
    prev: Option<Arc<Datatype>>,
}

fn round_snapshot(cells: &[RoundCell]) -> String {
    let mut parts = Vec::new();
    for cell in cells {
        let target = cell.target.read().unwrap();
        let datatype = target.get_type().expect("cell typed");
        let same_prev = cell
            .prev
            .as_ref()
            .map(|previous| Arc::ptr_eq(previous, &datatype))
            .unwrap_or(false) as u8;
        parts.push(format!(
            "{}:{}{}/same_prev={same_prev}",
            cell.label,
            metatype2string(datatype.get_metatype()),
            datatype.get_size()
        ));
    }
    parts.join(",")
}

fn main() {
    println!(
        "schema=1|fixture=INFERTYPES-SETTLE-0001|oracle=e40ed13014025f82488b1f8f7bca566894ac376b"
    );

    let type_factory = Arc::new(RwLock::new(TypeFactory::new_flavor(
        8,
        CoreTypeFlavor::Standalone,
    )));
    let (progress, progress_pointer) = {
        let mut factory = type_factory.write().unwrap();
        let long_type = factory
            .get_base(8, TypeMetatype::Int)
            .expect("canonical int8");
        let progress = Arc::new(Datatype::Struct(TypeStruct {
            base: TypeBase::new("ProgressData".into(), 16, TypeMetatype::Struct),
            fields: vec![
                TypeField {
                    name: "total".into(),
                    offset: 0,
                    type_ptr: long_type.clone(),
                },
                TypeField {
                    name: "prev".into(),
                    offset: 8,
                    type_ptr: long_type,
                },
            ],
        }));
        let pointer = factory.get_ptr(progress.clone());
        (progress, pointer)
    };
    let _ = &progress;

    let mut architecture = Architecture::new();
    architecture.archid = "x86:LE:64:default:gcc".to_string();
    architecture.set_types(type_factory.clone());
    let mut fd = Funcdata::new("infertypes_settle", Address::new(0x5000), 0x40);
    fd.vbank.set_type_factory(type_factory.clone());
    fd.set_arch(Arc::new(architecture));
    let block = fd.create_new_block();

    // An 8-byte int input feeding the float chain.
    let counter = fd.vbank.create_with_space(4, AddressSpace::Register, 0x100);
    let counter = fd.set_input_varnode(counter);

    // f2 = FLOAT_FLOAT2FLOAT(INT2FLOAT(counter) * const) : 4 -> 8 widening.
    let int2float = fd.new_op(1, Address::new(0x5000));
    fd.op_set_opcode(&int2float, OpCode::CPUI_FLOAT_INT2FLOAT);
    fd.op_set_input(&int2float, counter.clone(), 0);
    let float4 = fd.new_unique_out(4, &int2float);
    fd.op_insert_end(&int2float, &block);

    let fmul = fd.new_op(2, Address::new(0x5002));
    fd.op_set_opcode(&fmul, OpCode::CPUI_FLOAT_MULT);
    fd.op_set_input(&fmul, float4.clone(), 0);
    let scale = fd.new_constant(4, 0x3f800000);
    fd.op_set_input(&fmul, scale, 1);
    let product4 = fd.new_unique_out(4, &fmul);
    fd.op_insert_end(&fmul, &block);

    let f2f = fd.new_op(1, Address::new(0x5004));
    fd.op_set_opcode(&f2f, OpCode::CPUI_FLOAT_FLOAT2FLOAT);
    fd.op_set_input(&f2f, product4.clone(), 0);
    let double8 = fd.new_unique_out(8, &f2f);
    fd.op_insert_end(&f2f, &block);

    // Type-locked struct pointer driving LOAD/STORE pointer propagation.
    let pointer = fd.vbank.create_with_space(8, AddressSpace::Register, 0x180);
    let pointer = fd.set_input_varnode(pointer);
    pointer
        .write()
        .unwrap()
        .update_type_lock(progress_pointer.clone(), true, false);

    let load = fd.new_op(2, Address::new(0x5006));
    fd.op_set_opcode(&load, OpCode::CPUI_LOAD);
    let ram_id = fd.new_constant(4, AddressSpace::Ram.space_id() as u64);
    fd.op_set_input(&load, ram_id.clone(), 0);
    fd.op_set_input(&load, pointer.clone(), 1);
    let loaded8 = fd.new_unique_out(8, &load);
    fd.op_insert_end(&load, &block);

    let store = fd.new_op(3, Address::new(0x5008));
    fd.op_set_opcode(&store, OpCode::CPUI_STORE);
    fd.op_set_input(&store, ram_id, 0);
    fd.op_set_input(&store, pointer.clone(), 1);
    fd.op_set_input(&store, double8.clone(), 2);
    fd.op_insert_end(&store, &block);

    let mut cells: Vec<RoundCell> = Vec::new();
    for (label, target) in [
        ("counter", &counter),
        ("float4", &float4),
        ("product4", &product4),
        ("double8", &double8),
        ("pointer", &pointer),
        ("loaded8", &loaded8),
    ] {
        cells.push(RoundCell {
            label,
            target: target.clone(),
            prev: target.read().unwrap().get_type(),
        });
    }
    println!("pre|types={}", round_snapshot(&cells));

    fd.start_type_recovery();
    let mut action = ActionInferTypes::new();
    action.reset(&mut fd);
    for round in 1..=8 {
        let result = action
            .apply(&mut fd)
            .unwrap_or_else(|error| panic!("apply round {round} failed: {error}"));
        println!(
            "round{round}|return={result}|types={}",
            round_snapshot(&cells)
        );
        for cell in &mut cells {
            cell.prev = cell.target.read().unwrap().get_type();
        }
    }
    // Rust-only settle guard: local_count must stay below the 7-pass cap.
    assert!(
        action.local_count < 7,
        "ActionInferTypes did not settle: local_count={}",
        action.local_count
    );
}
